airbyte-integrations/connectors/source-attio/source_attio/source.py

```diff
@@ -0,0 +1,527 @@
+        stream_data = response.json()
+        num_items = len(stream_data["data"])
+
+        if num_items < self.limit:
+            return None
+
+        self.offset += num_items
+        return {"offset": self.offset}
+
+
+class Records(AttioStream):
+    """
+    This stream is responsible for handling records for a given object in the workspace.
+    Although all workspaces will have people and company objects enabled, there may also be other
+    standard (e.g. users, deals) or custom (e.g. projects, tasks, ...) objects enabled.
+    """
+
+    offset = 0
+
+    def __init__(self, object_slug: str, object_id: str, limit=DEFAULT_RECORDS_LIMIT, **kwargs):
+        self.object_slug = object_slug
+        self.object_id = object_id
+        self.limit = limit
+        super().__init__(**kwargs)
+
+    @property
+    def name(self) -> str:
+        return self.object_slug
+
+    @property
+    def http_method(self) -> str:
+        return "POST"
+
+    def path(self, **kwags) -> str:
+        """
+        The path to the query endpoint to get the list of records.
+        See: https://developers.attio.com/reference/post_v2-objects-object-records-query
+        """
+        return f"objects/{self.object_id}/records/query"
+
+    def parse_response(self, response: requests.Response, **kwargs) -> Iterable[Mapping]:
+        """
+        Conforms to the schema returned by get_json_schema
+        """
+        object_id = self.object_id
+        object_slug = self.object_slug
+
+        attributes_stream = ObjectAttributes(authenticator=self.authenticator, object_id=object_id, object_slug=object_slug)
+        # Attributes stream is cached
+        attributes = list(attributes_stream.read_records(sync_mode=SyncMode.full_refresh))
+
+        response_json = response.json()
+        for record in response_json["data"]:
+            # Flatten ID columns
+            record["workspace_id"] = record["id"]["workspace_id"]
+            record["object_id"] = record["id"]["object_id"]
+            record["record_id"] = record["id"]["record_id"]
+            del record["id"]
+
+            # Iterate and output values in conformance with the schema generated by get_json_schema.
+            # values is a dict from attribute api_slug to an array of values.
+            values = record["values"]
+
+            del record["values"]
+
+            for attribute in attributes:
+                t = attribute["type"]
+
+                if t == "interaction":
+                    continue
+
+                slug = attribute["api_slug"]
+                is_multi = attribute["is_multiselect"]
+
+                vs = values[slug]
+
+                if t == "text" and not is_multi:
+                    record[slug] = None if len(vs) == 0 else vs[0]["value"]
+                elif t == "number" and not is_multi:
+                    record[slug] = None if len(vs) == 0 else vs[0]["value"]
```

Comment:
I feel like we could make this a bit more config driven (and combine with the dynamic json schema logic) with a some helper functions and a config map rather than a really long `if/elif`

```python
### Unpackers
def simple_value_unpacker(field_name: str):
    def unpack(value, slug):
        return {
            slug: value[field_name]
        }

    return unpack

def record_reference_unpacker(value: Mapping, slug: str):
    return {
        slug: {
            "target_object": value["target_object"],
            "target_record_id": value["target_record_id"],
        }
    }

def location_unpacker(value: Mapping, slug: str):
    return {
        slug + "_line_1": value["line_1"],
        slug + "_line_2": value["line_2"],
        # ...
    }

### JSONSchema generators
def simple_attribute_json_schema(value_json_schema: Mapping[str, str]):
    def attribute_json_schema(slug: str):
        return {
            slug: value_json_schema
        }

    return attribute_json_schema
    
def location_attribute_json_schema(slug: str):
    return {
        slug + "_line_1": {"type": "string"},
        slug + "_line_2": {"type": "string"}
    }

ATTRIBUTE_TYPE_CONFIG = {
    'text': {
        'supports_multi': False,
        'unpack_value': simple_value_unpacker("value"),
        'json_schema_properties': simple_attribute_json_schema({ "type": "text" })
    },
    'currency': {
        'supports_multi': False,
        'unpack_value': simple_value_unpacker("currency_value"),
        'json_schema_properties': simple_attribute_json_schema({ "type": "number" })
    },
    # ...
    'location': {
        'supports_multi': False,
        'unpack_value': location_unpacker,
        'json_schema_properties': location_attribute_json_schema
    },
    'record-reference': {
        'supports_multi': True,
        'unpack_value': record_reference_unpacker,
        'json_schema_properties': simple_attribute_json_schema({
            'referenced_actor_id': {"type": "string"},
            'referenced_actor_type': {"type": "string"},
        })
    },
}
```

We can then extend this a bit easier with other value types and the actual code to do the unwrapping can be a bit more generic, auto handling wrapping the json schema type with `{"type": "array"}` / iterating over the values when  `is_multiselect == True` (and `supports_multi == True` in the config)

The return values of these function can then be added to `record` or schema `properties` using `dict.update`

wdyt?

Reply:
@mulholo are you making these changes or sticking with the current implementation? 

We were about to start using this source, but if you're making the changes, we'll hold off for those.

Reply:
We're going to make a few tweaks to the behaviour of this source after running through some internal testing. Hope to get done end of week/start of next week. This will mostly be around the representation of values in the schema, likely with some internal refactors to accompany.